From: Andrea Pesaresi <andreapesaresi82@gmail.com>
Date: Sat, 17 Oct 2026 20:30:48 +0200
Subject: [PATCH] adduser: read the password from a pipe

When stdin is not a terminal, tcgetattr() fails and term is left
uninitialized. raw_term.c_cc[VERASE] then holds a random byte, which is
handled as backspace whenever it shows up in the password, so a password
piped to ksmbd.adduser can be stored altered without any error.

Only switch the terminal to raw mode, handle the erase character and
restore the settings when stdin is a terminal. This lets programs pass
the password on stdin instead of putting it on the command line.

Signed-off-by: Andrea Pesaresi <andreapesaresi82@gmail.com>
---
 adduser/user_admin.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/adduser/user_admin.c b/adduser/user_admin.c
index 59c1336..b7be95c 100644
--- a/adduser/user_admin.c
+++ b/adduser/user_admin.c
@@ -29,14 +29,23 @@ static void __prompt_password_stdin(char *password, size_t *sz)
 	struct termios term, raw_term;
 	char buf[LINE_MAX];
 	size_t buflen;
+	int is_tty;
 
 	pr_info("Prompting for password\n");
 
-	tcgetattr(STDIN_FILENO, &term);
+	/*
+	 * stdin can be a pipe, e.g. when a program passes the password
+	 * without putting it on the command line. There is no terminal
+	 * to configure then, and term would be left uninitialized.
+	 */
+	memset(&term, 0, sizeof(term));
+	is_tty = !tcgetattr(STDIN_FILENO, &term);
 	raw_term = term;
 
-	raw_term.c_lflag &= ~(ECHO | ICANON);
-	tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw_term);
+	if (is_tty) {
+		raw_term.c_lflag &= ~(ECHO | ICANON);
+		tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw_term);
+	}
 
 	for (*sz = (size_t)-1,
 	     buflen = 0, password[buflen] = buf[buflen] = 0x00;;) {
@@ -82,7 +91,7 @@ static void __prompt_password_stdin(char *password, size_t *sz)
 			break;
 		}
 
-		if (c == raw_term.c_cc[VERASE]) {
+		if (is_tty && c == raw_term.c_cc[VERASE]) {
 			char *u8c;
 			size_t clen;
 
@@ -103,7 +112,8 @@ static void __prompt_password_stdin(char *password, size_t *sz)
 
 	printf("\eD" "\r" "\e[2K");
 
-	tcsetattr(STDIN_FILENO, TCSAFLUSH, &term);
+	if (is_tty)
+		tcsetattr(STDIN_FILENO, TCSAFLUSH, &term);
 }
 
 static int __is_valid_password_len(size_t len)
